### ISSUE TITLE
turbolinks:load fires twice // turbolinks-visit-control

### ISSUE THREAD

**Issue by grund3g** [turbolinks/turbolinks#441]
I have the problem that the `turbolinks:load` event is (obviously) fired twice when i use `<meta name="turbolinks-visit-control" content="reload">` on a page. 

As I also send google analytics pageviews with the `turbolinks:load` event, those events are also sent twice. 

Is there any way to tell if this event was caused by `turbolinks-visit-control`?

**Comment by zkingdesign** [comment]
Not sure if it's the same issue or not but my javascript calls seem to be stacking.

I have the following code in my app.js amongst many others:
```javascript
$(document).on("turbolinks:load", function(){
    console.log(event);
})
```

And when browsing through pages I get this in my console, http://prntscr.com/mgy3i9 , so with each page visit the stack increases by +1.

What am I doing wrong?

**Comment by grund3g** [comment]
@zkingdesign I don't actually think it's the same problem. On your site it looks like you are setting up the event listener on each page change. Are you sure that you are loading your script in your `<head>` and not in your `<body>` ? (see the [docs](https://github.com/turbolinks/turbolinks#loading-your-applications-javascript-bundle))

**Comment by zkingdesign** [comment]
@grund3g yes I am loading it in the <head> of the site, see here http://prntscr.com/mgz7c8

**Comment by grund3g** [comment]
@zkingdesign No, I meant the script where you are setting up your event listener: 
```
$(document).on("turbolinks:load", function(){
    console.log(event);
})
```

**Comment by domchristie** [comment]
> Is there any way to tell if this event was caused by `turbolinks-visit-control`?

@grund3g there is currently no way to detect whether a page load event was caused by the page being invalidated i.e. reloaded after `turbolinks-visit-control` (reload) or tracked asset change.

However, it is possible to patch the `reload` method to track when a reload was triggered, and then decide how to handle the subsequent load event. For example, the following snippet sets up a place to track `turbolinks:load` events when they were _not_ triggered by a reload:

```js
;(function () {
  var PAGE_INVALIDATED = 'page-invalidated'

  // Patch the reload method to flag that the following page load originated from the page being invalidated
  Turbolinks.BrowserAdapter.prototype.reload = function () {
    sessionStorage.setItem(PAGE_INVALIDATED, 'true')
    location.reload()
  }
  
  addEventListener('turbolinks:load', function (event) {
    if (!JSON.parse(sessionStorage.getItem(PAGE_INVALIDATED))) {
      // Track (non-reloaded) visit
    }
    sessionStorage.removeItem(PAGE_INVALIDATED)
  })
})()
```

A couple of caveats: I'm not sure how things like "time on page" analytics work, so metrics like this could be affected 🤷‍♂️; secondly this patches undocumented API and so may change in future versions; and finally this won't work when using the iOS/android adapters.

Detecting how a page has been loaded seems like it might be a useful feature\*, so I suppose this could be a candidate for an enhancement.

Hope that helps!

---
\*  See issue: https://github.com/turbolinks/turbolinks/issues/115 and related PR: https://github.com/turbolinks/turbolinks/pull/286, as well as https://github.com/turbolinks/turbolinks/issues/181#issuecomment-423736911 or https://github.com/turbolinks/turbolinks/pull/420 which use similar techniques to the above.

**Comment by grund3g** [comment]
@domchristie Excellent, thank you! That's what I was looking for. As I don't use any of the analytics, that shouldn't be a problem for me. 

But yes, that would certainly be a good feature for the future!
